awx/main/scheduler/task_manager_models.py

```diff
@@ -0,0 +1,110 @@
+        self.capacity = capacity
+        self.hostname = hostname
+
+
+class TaskManagerInstances:
+    def __init__(self, active_tasks=[]):
+        instances = Instance.objects.filter(hostname__isnull=False, enabled=True).exclude(node_type='hop').only('node_type', 'capacity', 'hostname', 'enabled')
+        self._remaining_capacity_updated = False
+        instance_list = [
+            TaskManagerInstance(
+                obj=instance,
+                node_type=instance.node_type,
+                capacity=instance.capacity,
+                hostname=instance.hostname,
+            )
+            for instance in instances
+        ]
+        self._update_remaining_capacity(active_tasks)
+        self.instances_by_hostname = {i.hostname: i for i in instance_list}
+
+    def _update_remaining_capacity(self, tasks):
+        """Takes a list of tasks and updates the remaining capacity on our TaskManagerInstance list.
+
+        Computes remaining capacity for all the instances based on currently running and waiting tasks.
+        """
+        if self._remaining_capacity_updated:
+            # Only want to do this once
+            return
+        for task in tasks:
+            if task.status not in ['waiting', 'running']:
+                continue
+            control_instance = self.instances_by_hostname.get(task.controller_node, '')
+            execution_instance = self.instances_by_hostname.get(task.execution_node, '')
+            if execution_instance and execution_instance.node_type in ('hybrid', 'execution'):
+                self.instances_by_hostname[task.execution_node].remaining_capacity -= task.task_impact
+            if control_instance and control_instance.node_type in ('hybrid', 'control'):
+                self.instances_by_hostname[task.controller_node].remaining_capacity -= settings.AWX_CONTROL_NODE_TASK_IMPACT
+
+    def fit_task_to_most_remaining_capacity_instance(self, task, impact=None, capacity_type=None, add_hybrid_control_cost=False):
+        impact = impact if impact else task.task_impact
+        capacity_type = capacity_type if capacity_type else task.capacity_type
+        instance_most_capacity = None
+        most_remaining_capacity = -1
+        for i in self.instances_by_hostname.values():
+            if i.node_type not in (capacity_type, 'hybrid'):
+                continue
+            would_be_remaining = i.remaining_capacity - impact
+            # hybrid nodes _always_ control their own tasks
+            if add_hybrid_control_cost and i.node_type == 'hybrid':
+                would_be_remaining -= settings.AWX_CONTROL_NODE_TASK_IMPACT
+            if would_be_remaining >= 0 and (instance_most_capacity is None or would_be_remaining > most_remaining_capacity):
+                instance_most_capacity = i
+                most_remaining_capacity = would_be_remaining
+        return instance_most_capacity
+
+    def find_largest_idle_instance(self, capacity_type='execution'):
```

Comment:
TODO need to accept the instance group name we want it to be in or move this to the TaskManagerInstanceGroups class